### PR TITLE
Feature lwdev 5149 layer rule

### PR DESCRIPTION
**Pull request by aushakov**

### CODE REVIEW

File: src/Lykke.Service.ClientAssetRule/Settings/Clients/AssetsServiceClient.cs [hunk 1]
```diff
@@ -0,0 +1,7 @@
+﻿namespace Lykke.Service.ClientAssetRule.Settings.Clients
+{
+    public class AssetsServiceClientSettings
+    {
+        public string ServiceUrl { get; set; }
```
Comment (samodovdi) [inline src/Lykke.Service.ClientAssetRule/Settings/Clients/AssetsServiceClient.cs]: [HttpCheck] attribute can be added

File: src/Lykke.Service.ClientAssetRule/Settings/ServiceSettings/Rabbit/RegulationQueue.cs [hunk 2]
```diff
@@ -1,4 +1,4 @@
-﻿namespace Lykke.Service.ClientAssetRule.Core.Settings.ServiceSettings
+﻿namespace Lykke.Service.ClientAssetRule.Settings.ServiceSettings.Rabbit
 {
     public class RegulationQueue
```
Comment (samodovdi) [inline src/Lykke.Service.ClientAssetRule/Settings/ServiceSettings/Rabbit/RegulationQueue.cs]: [AmqpCheck] attribute can be added on ConnectionString

**Review by samodovdi — APPROVED** [review]

File: src/Lykke.Service.ClientAssetRule/Settings/Clients/AssetsServiceClient.cs (hunk 1 above)
Comment (aushakov) [reply src/Lykke.Service.ClientAssetRule/Settings/Clients/AssetsServiceClient.cs]: It's an external service.

File: client/Lykke.Service.ClientAssetRule.Client/Models/AssetConditionLayerRuleModel.cs [hunk 3]
```diff
@@ -0,0 +1,13 @@
+﻿using System.Collections.Generic;
+
+namespace Lykke.Service.ClientAssetRule.Client.Models
+{
+    public class AssetConditionLayerRuleModel
```
Comment (MikhailGerasimov) [inline client/Lykke.Service.ClientAssetRule.Client/Models/AssetConditionLayerRuleModel.cs]: No description for the public model

File: src/Lykke.Service.ClientAssetRule.AzureRepositories/AssetConditionLayer/AssetConditionLayerRuleRepository.cs [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+    public class AssetConditionLayerRuleRepository : IAssetConditionLayerRuleRepository
+    {
+        private readonly INoSQLTableStorage<AssetConditionLayerRuleEntity> _storage;
+
+        public AssetConditionLayerRuleRepository(INoSQLTableStorage<AssetConditionLayerRuleEntity> storage)
+        {
+            _storage = storage;
+        }
+
+        public async Task<IList<IAssetConditionLayerRule>> GetAsync()
```
Comment (MikhailGerasimov) [inline src/Lykke.Service.ClientAssetRule.AzureRepositories/AssetConditionLayer/AssetConditionLayerRuleRepository.cs]: Why not IReadOnlyList?

File: src/Lykke.Service.ClientAssetRule/AutofacModule.cs [hunk 5]
```diff
@@ -0,0 +1,39 @@
...
+            _log = log;
+        }
+
+        protected override void Load(ContainerBuilder builder)
+        {
+            builder.RegisterInstance(_log)
+                .As<ILog>()
+                .SingleInstance();
+
+            builder.RegisterInstance<IAssetsService>(new AssetsService(new Uri(_settings.AssetsServiceClient.ServiceUrl)));
```
Comment (MikhailGerasimov) [inline src/Lykke.Service.ClientAssetRule/AutofacModule.cs:28]: AssetService has its own extension method for registration?

File: src/Lykke.Service.ClientAssetRule.Services/AssetConditionLayerRuleService.cs [hunk 6]
```diff
@@ -0,0 +1,109 @@
...
+        {
+            return await _assetConditionLayerRuleRepository.GetAsync(regulationId);
+        }
+
+        public async Task SetAsync(string clientId, IEnumerable<string> regulations)
+        {
+            IList<IAssetConditionLayerRule> rules = await _assetConditionLayerRuleRepository.GetAsync();
+
+            List<string> allowed = rules
+                .Where(o => regulations.Contains(o.RegulationId))
```
Comment (MikhailGerasimov) [inline src/Lykke.Service.ClientAssetRule.Services/AssetConditionLayerRuleService.cs]: Performance issues can be here the number of operations is rules * regulations

File: src/Lykke.Service.ClientAssetRule.Services/AssetConditionLayerRuleService.cs [hunk 7]
```diff
@@ -0,0 +1,109 @@
...
+        public async Task<IAssetConditionLayerRule> GetAsync(string regulationId)
+        {
+            return await _assetConditionLayerRuleRepository.GetAsync(regulationId);
+        }
+
+        public async Task SetAsync(string clientId, IEnumerable<string> regulations)
+        {
+            IList<IAssetConditionLayerRule> rules = await _assetConditionLayerRuleRepository.GetAsync();
+
+            List<string> allowed = rules
```
Comment (MikhailGerasimov) [inline src/Lykke.Service.ClientAssetRule.Services/AssetConditionLayerRuleService.cs]: HashSet would be better. https://github.com/LykkeCity/Lykke.Service.ClientAssetRule/pull/5/files#diff-44b0aadf07a12762f1652a019a22befaR54

File: src/Lykke.Service.ClientAssetRule.Services/AssetGroupRuleService.cs [hunk 8]
```diff
@@ -0,0 +1,121 @@
...
+
+                foreach (IAssetGroupRule rule in rules)
+                {
+                    allowed.AddRange(rule.AllowedAssetGroups);
+                    declined.AddRange(rule.DeclinedAssetGroups);
+                }
+            }
+
+            declined = declined.Distinct()
+                .ToList();
```
Comment (MikhailGerasimov) [inline src/Lykke.Service.ClientAssetRule.Services/AssetGroupRuleService.cs]: ToHashSet better

File: tests/Lykke.Service.ClientAssetRule.AzureRepositories.Tests/AssetConditionLayerRuleRepositoryTests.cs [hunk 9]
```diff
@@ -0,0 +1,14 @@
+﻿using Microsoft.VisualStudio.TestTools.UnitTesting;
+
+namespace Lykke.Service.ClientAssetRule.AzureRepositories.Tests
+{
+    [TestClass]
+    public class AssetConditionLayerRuleRepositoryTests
```
Comment (MikhailGerasimov) [inline tests/Lykke.Service.ClientAssetRule.AzureRepositories.Tests/AssetConditionLayerRuleRepositoryTests.cs:6]: What for this test?

File: tests/Lykke.Service.ClientAssetRule.Tests/AssetConditionLayerRuleControllerTests.cs [hunk 10]
```diff
@@ -0,0 +1,14 @@
+﻿using Microsoft.VisualStudio.TestTools.UnitTesting;
+
+namespace Lykke.Service.ClientAssetRule.Tests
+{
+    [TestClass]
+    public class AssetConditionLayerRuleControllerTests
+    {
+        [TestMethod]
+        public void Test_OK()
```
Comment (MikhailGerasimov) [inline tests/Lykke.Service.ClientAssetRule.Tests/AssetConditionLayerRuleControllerTests.cs:9]: What for this test?

Comment (aushakov) [reply tests/Lykke.Service.ClientAssetRule.Tests/AssetConditionLayerRuleControllerTests.cs:9]: Now it's only to test AutoMapper config

File: tests/Lykke.Service.ClientAssetRule.AzureRepositories.Tests/AssetConditionLayerRuleRepositoryTests.cs (hunk 9 above)
Comment (aushakov) [reply tests/Lykke.Service.ClientAssetRule.AzureRepositories.Tests/AssetConditionLayerRuleRepositoryTests.cs:6]: Now it's only to test AutoMapper config

File: src/Lykke.Service.ClientAssetRule/AutofacModule.cs (hunk 5 above)
Comment (aushakov) [reply src/Lykke.Service.ClientAssetRule/AutofacModule.cs:28]: No need all services registration in the ServiceCollectionExtensions

**Review by MikhailGerasimov — APPROVED** [review]

**Pull request merged**